synapse/rest/client/room.py

```diff
@@ -556,13 +622,18 @@ class RoomMessageListRestServlet(RestServlet):
     def __init__(self, hs: "HomeServer"):
         super().__init__()
         self._hs = hs
+        self.clock = hs.get_clock()
         self.pagination_handler = hs.get_pagination_handler()
         self.auth = hs.get_auth()
         self.store = hs.get_datastores().main
 
     async def on_GET(
         self, request: SynapseRequest, room_id: str
     ) -> Tuple[int, JsonDict]:
+        processing_start_time = self.clock.time_msec()
+        # Fire and forget and hope that we get a result by the end.
+        room_member_count_co = self.store.get_number_joined_users_in_room(room_id)
```

Comment:
This won't fire and forget anything IIRC — creating a coroutines won't actually make it work; you need to pass it off to something (event loop or 'Reactor' in Twisted parlance) to make it work convert; in Twisted you do this by converting it to a deferred with `ensureDeferred`. However there are log context implications; consider asking someone properly in #synapse-dev as I'm not confident I can tell you what's what.

Also see https://matrix-org.github.io/synapse/latest/log_contexts.html?highlight=log%20context#log-contexts in case your case is an easy case. :-)

Reply:
As you suspected, it looks like it does have log context implications. When using `ensureDeferred`, I am seeing the tracing lose its contexts (we have a `scopecontextmanager.py` which piggybacks on the log contexts).

```
2022-08-17 16:33:57,435 - synapse.logging.opentracing - 404 - ERROR - sentinel - There was no active span when trying to set a tag. Did you forget to start one or did a context slip?
Stack (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/eric/Documents/github/element/synapse/synapse/app/homeserver.py", line 422, in main
    run(hs)
  File "/Users/eric/Documents/github/element/synapse/synapse/app/homeserver.py", line 401, in run
    _base.start_reactor(
  File "/Users/eric/Documents/github/element/synapse/synapse/app/_base.py", line 191, in start_reactor
    run()
  File "/Users/eric/Documents/github/element/synapse/synapse/app/_base.py", line 173, in run
    run_command()
  File "/Users/eric/Documents/github/element/synapse/synapse/app/_base.py", line 148, in <lambda>
    run_command: Callable[[], None] = lambda: reactor.run(),
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/base.py", line 1315, in run
    self.mainLoop()
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/base.py", line 1325, in mainLoop
    reactorBaseSelf.runUntilCurrent()
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/base.py", line 964, in runUntilCurrent
    f(*a, **kw)
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 661, in callback
    self._startRunCallbacks(result)
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 763, in _startRunCallbacks
    self._runCallbacks()
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 857, in _runCallbacks
    current.result = callback(  # type: ignore[misc]
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1750, in gotResult
    current_context.run(_inlineCallbacks, r, gen, status)
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1663, in _inlineCallbacks
    status.deferred.callback(getattr(e, "value", None))
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 661, in callback
    self._startRunCallbacks(result)
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 763, in _startRunCallbacks
    self._runCallbacks()
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 857, in _runCallbacks
    current.result = callback(  # type: ignore[misc]
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 661, in callback
    self._startRunCallbacks(result)
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 763, in _startRunCallbacks
    self._runCallbacks()
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 857, in _runCallbacks
    current.result = callback(  # type: ignore[misc]
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1750, in gotResult
    current_context.run(_inlineCallbacks, r, gen, status)
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1660, in _inlineCallbacks
    result = current_context.run(gen.send, result)
  File "/Users/eric/Documents/github/element/synapse/synapse/http/server.py", line 290, in wrapped_async_request_handler
    await h(self, request)
  File "/Users/eric/Documents/github/element/synapse/synapse/http/server.py", line 368, in _async_render_wrapper
    callback_return = await self._async_render(request)
  File "/Users/eric/Documents/github/element/synapse/synapse/http/server.py", line 574, in _async_render
    callback_return = await raw_callback_return
  File "/Users/eric/Documents/github/element/synapse/synapse/rest/client/room.py", line 662, in on_GET
    msgs = await self.pagination_handler.get_messages(
  File "/Users/eric/Documents/github/element/synapse/synapse/logging/opentracing.py", line 876, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/Users/eric/Documents/github/element/synapse/synapse/handlers/pagination.py", line 499, in get_messages
    await self.hs.get_federation_handler().maybe_backfill(
  File "/Users/eric/Documents/github/element/synapse/synapse/logging/opentracing.py", line 876, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/Users/eric/Documents/github/element/synapse/synapse/handlers/federation.py", line 225, in maybe_backfill
    return await self._maybe_backfill_inner(
  File "/Users/eric/Documents/github/element/synapse/synapse/handlers/federation.py", line 255, in _maybe_backfill_inner
    for event_id, depth in await self.store.get_oldest_event_ids_with_depth_in_room(
  File "/Users/eric/Documents/github/element/synapse/synapse/logging/opentracing.py", line 876, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/Users/eric/Documents/github/element/synapse/synapse/logging/opentracing.py", line 875, in _wrapper
    with wrapping_logic(func, *args, **kwargs):
  File "/usr/local/Cellar/python@3.9/3.9.13_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/contextlib.py", line 119, in __enter__
    return next(self.gen)
  File "/Users/eric/Documents/github/element/synapse/synapse/logging/opentracing.py", line 984, in _wrapping_logic
    set_tag(SynapseTags.FUNC_ARGS, str(args[len(argspec.args) :]))  # type: ignore[index]
  File "/Users/eric/Documents/github/element/synapse/synapse/logging/opentracing.py", line 404, in ensure_active_span_inner_2
    logger.error(
```

Reply:
@DMRobertson [suggested](https://matrix.to/#/!vcyiEtMVHIhWXcJAfl:sw1v.org/$54v1XiUzchCDXy0eptLYxInzBgt3GK4sQuKMClTpN1Y?via=matrix.org&via=element.io&via=beeper.com) using `run_as_background_process` which does seem to work 😀

Reply:
from discussion in #synapse-dev this morning, `run_in_background` is what you want since you don't want it to be tracked as part of a background process's metrics :)

Reply:
Discussion from `#synapse-dev` for reference:

> > <@dmrobertson:matrix.org> (surprised that's in synapse.metrics)
>
> that's mostly because `run_as_background_process` isn't actually what you want, because that's all about reporting the CPU and DB usage metrics in the \"background jobs\" section of the grafana dashboard. If what you're looking for is doing things in parallel, and you're actually going to wait for the thing to complete at the end, then in theory what you want to do is to `Deferred.fromCoroutine(get_number_joined_users_in_room())` (which gives you a `Deferred` which is much like a javascript Promise;  it will run in the background and you can await it later).
>
> *However* our logcontexts mess that up so actually you need more wrapping, so in practice the thing you *actually* want here is `synapse.logging.context.run_in_background(get_number_joined_users_in_room())`
>
> and whether you use `run_as_background_process` or `run_in_background`, when you finally do the await at the end, you need to `await make_deferred_yieldable(deferred)`, otherwise again logcontexts will get messed up if `deferred` isn't complete before you await it.
>
> once more: the sooner we switch to proper contextvars rather than having to worry about the logcontext magic, the better
>
> *-- @richvdh, https://matrix.to/#/!vcyiEtMVHIhWXcJAfl:sw1v.org/$UdCC3RUwlnyyf1gJnxxXdXNIN7syssmJve9RzftD180?via=matrix.org&via=element.io&via=beeper.com*

Reply:
Updated ✅

Reply:
There is a documentation section about fire-and-forget: https://github.com/matrix-org/synapse/blob/b6baa46db078c3ef9e6c5751bccb8d2e1c5c5402/docs/log_contexts.md#fire-and-forget